Improved support for Bluefish hardware + new features

jamesdwise opened the pull request:
This updates and improves the bluefish module and hardware support, namely:
- Change the API interface to use the more recent BlueVelvetC
- Change the way the module links against the header and libs.
- Add support for the latest Bluefish hardware
- Add support for new features.

This work is conducted on behalf of Bluefish44 to ensure their 
products are correctly integrated and supported with CasparCG

Change the API interface to use the more recent BlueVelvetC:
	Created a BVCWrapper class that uses the Latest BlueVelvetC API. 
	This allows for caspar to support the most recent hardware and features.
	It also simplifies the build and link process.	

Change the way the module links against the header:
	The new BVCWrapper class just includes headers and attempts to load functions pointers from a DLL if found.
	This has allowed us to remove all files from the dependencies64/bluefish dir, reducing project size and dependencies.
	Created a new bluefish/interop dir that stores the updated and required header files.
	
Add support for the latest Bluefish hardware
	Latest verison of driver_p.h file has been used, this includes supoprt for the most recent Bluefish hardware
	There exist mutliple version of recent hardware that vary by what is supported, eg SDI i/o count, and presence of 		onboard scaler hardware or not.  
	Code has been added to support all these different configurations and variations,
 	and includes improved routing and configuration, required to support new features, below..

Add support for new features.
	The bluefish module now supports additional features: 
		Use of onboard embedded audio, 
		Selection of what output channel and type to use, ( ie. 4224, vs 422 )
		Support for onboard keyer, if supported in hardware.
	I have also updated the casparcg.config XML file, 
	Added example  of how to use the new features in the comments section.

Outstanding features and concerns.
1) The core BVCwrapper class can be used across all platforms, 
	but currently only windows is enabled and working in the code.
	Loading of function pointers still needs to be added for OSX and Linux.
	Other very minot OS specific issues will need to be addressed inside the BVCWrapper class.
2) CMakefile has been modified to ensure that bluefish module only gets built on win.
3) CMakefiles have been updated to remove the bluefish64 dir from the dependcies.
4) Further work will involve adding a Bluefish Consumer, and making the current module supported across all platforms.

hummelstrand commented:
Great to see Bluefish Tehnologies involved, thanks James!

HellGore commented on modules/bluefish/CMakeLists.txt:
```diff
@@ -3,18 +3,18 @@ project (bluefish)
+		interop/BlueDriver_p.h
+		interop/BlueHancUtils.h
+		interop/BlueVelvetC.h
+		interop/BlueVelvetCUtils.h
```
I think these should still have been under dependencies64/bluefish/include as before since they are part of your SDK not really part of CasparCG. What do you think?

jamesdwise replied:
I was simply copying what was done with the blackmagic module - Which may or may not have been a good idea.  I am happy to move them to dependencies64/bluefish/include, As i personally feel that is probably the correct location for the files.

Personally i tend to agree that

HellGore commented on modules/bluefish/bluefish.cpp:
```diff
@@ -78,21 +75,42 @@ std::vector<std::wstring> device_list()
 void init(core::module_dependencies dependencies)
 {
+
+	bool bCanBeConsumer = false;
```
Please don't use camel case (only lower case with underscores) and also don't use hungarian notation.

HellGore commented on modules/bluefish/bluefish.cpp:
```diff
@@ -78,21 +75,42 @@ std::vector<std::wstring> device_list()
-	dependencies.consumer_registry->register_consumer_factory(L"Bluefish Consumer", create_consumer, describe_consumer);
-	dependencies.consumer_registry->register_preconfigured_consumer_factory(L"bluefish", create_preconfigured_consumer);
-	dependencies.system_info_provider_repo->register_system_info_provider([](boost::property_tree::wptree& info)
+	if (bCanBeConsumer)
```
It is better if the consumer is registered even though there are no capable Bluefish cards in the computer, so for example

```
HELP CONSUMERS
```

...always includes Bluefish Consumer even though it is not possible to instantiate one based on the current system configuration.

HellGore commented on modules/bluefish/bluefish.cpp:
```diff
@@ -78,21 +75,42 @@ std::vector<std::wstring> device_list()
+		dependencies.consumer_registry->register_preconfigured_consumer_factory(L"bluefish", create_preconfigured_consumer);
+		dependencies.system_info_provider_repo->register_system_info_provider([](boost::property_tree::wptree& info)
+		{
+			info.add(L"system.bluefish.version", version());
```
This is also an example of something that should be included even though Bluefish drivers are not installed on a system so that the startup log will show `bluefish.version Not found`

HellGore commented on modules/bluefish/util/blue_velvet.h:
```diff
@@ -33,25 +32,213 @@
 }
 
+template<typename T>
+int get_card_property(T& pSdk, ULONG prop, ULONG& value)
```
Change to

```
int get_card_property(T& pSdk, int prop, unsigned int& value)
```

so the unnecessary lines below can be removed.

HellGore commented on modules/bluefish/util/blue_velvet.h:
```diff
@@ -33,25 +32,213 @@
-EVideoMode get_video_mode(CBlueVelvet4& blue, const core::video_format_desc& format_desc);
 
 template<typename T>
 int set_card_property(T& pSdk, ULONG prop, ULONG value)
```
```
int set_card_property(T& pSdk, int prop, unsigned int value)
```

...is better.

HellGore commented on modules/CMakeLists.txt:
```diff
@@ -20,4 +19,5 @@ if (MSVC)
 endif ()
 
 add_subdirectory(image)
+add_subdirectory(reroute)
```
What was the reason for this initialization order move?

jamesdwise replied:
Ahh no reason, I suspect this was left over from my (experimental) work in adding the producer code,   Will revert.

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -56,10 +56,78 @@
+{
+	switch (streamid)
+	{
+	case bluefish_hardware_output_channel::hardware_4224_channel_a:	channelSrcElement = EPOCH_SRC_OUTPUT_MEM_INTERFACE_CHA;
```
There should be a newline directly after `case bluefish_hardware_output_channel::hardware_4224_channel_a:`

HellGore commented on modules/bluefish/util/blue_velvet.h:
```diff
@@ -33,25 +32,213 @@
+#endif
+
+//BlueVelvetC function pointers members
+	pFunc_bfcGetVersion bfcGetVersion;
```
All of these function pointers should be initialized to `nullptr`. Especially `bfcFactory`.

HellGore commented on modules/bluefish/util/blue_velvet.cpp:
```diff
@@ -112,100 +276,274 @@ EVideoMode vid_fmt_from_video_format(const core::video_format& fmt)
 {
-	if(!BlueVelvetFactory4 || !encode_hanc_frame)
+	auto pWrap = new BvcWrapper();
+	if (!pWrap->IsBvcValid())
```
This should be thrown in the constructor of `BvcWrapper` instead. In that way a `BvcWrapper` instance can never be "not valid"

HellGore commented on modules/bluefish/util/blue_velvet.cpp:
```diff
@@ -20,67 +20,231 @@
+		hModule = LoadLibraryExA("BlueVelvetC64_d.dll", NULL, 0);
 #else
-	std::string module_str = "BlueHancUtils64.dll";
+		hModule = LoadLibraryExA("BlueVelvetC64.dll", NULL, 0);
```
RAII should be used by changing the type of `BvcWrapper::hModule` to `std::shared_ptr<void>` using `FreeLibrary` as the custom deleter.

HellGore commented on modules/bluefish/util/blue_velvet.cpp:
```diff
@@ -20,67 +20,231 @@
+#endif
 
-#include <core/video_format.h>
+#if defined(_WIN32)
```
Throw exception using `CASPAR_THROW_EXCEPTION` instead of returning `false`.

HellGore commented on modules/bluefish/util/blue_velvet.cpp:
```diff
@@ -20,67 +20,231 @@
+		bvc = nullptr;
+		hModule = nullptr;
+		InitFunctionsPointers();
+		if (bfcFactory)
```
Currently this is undefined behaviour since `bfcFactory` is not initialized to `nullptr`. This should not be needed anyway if my other suggestions about throwing an exception during `InitFunctionsPointers()`

HellGore commented on modules/bluefish/util/blue_velvet.cpp:
```diff
@@ -20,67 +20,231 @@
+		hModule = nullptr;
+		InitFunctionsPointers();
+		if (bfcFactory)
+			bvc = bfcFactory();
```
RAII should be used here as well with `bfcDestroy` as custom deleter of the `shared_ptr` constructor.

HellGore commented on modules/bluefish/util/blue_velvet.cpp:
```diff
@@ -20,67 +20,231 @@
+			bvc = bfcFactory();
+	}
+
+	BvcWrapper::~BvcWrapper()
```
With my other suggestions regarding RAII and `hModule` and `bvc` this destructor should be removed completely.

HellGore commented on modules/bluefish/util/blue_velvet.h:
```diff
@@ -33,25 +32,213 @@
+
+private:
+	bool				InitFunctionsPointers();
+	BLUEVELVETC_HANDLE	bvc;
```
`bvc` Needs to be after `hModule` in order for the RAII I have talked about in my other comments to be done in the correct order.

HellGore commented on modules/bluefish/util/blue_velvet.cpp:
```diff
@@ -20,67 +20,231 @@
+		return bRes;
+	}
+
+	bool BvcWrapper::IsBvcValid()
```
Should be removed completely after the changes in my other comments, `bvc` should never be `nullptr` if the `BvcWrapper` has been contstructed.

HellGore commented:
I have now tested the code with a "BlueFirmwareUpdate_Neutron_3o_V012.exe" firmware for 3 SDI out, and I get very bad performance with uneven channel pace. My config is:

```
  <channels>
    <channel>
      <video-mode>1080i5000</video-mode>
      <consumers>
        <bluefish>
           <device>1</device>
           <sdi-stream>blue_422_channela</sdi-stream>
           <embedded-audio>true</embedded-audio>
        </bluefish>
      </consumers>
    </channel>
    <channel>
      <video-mode>1080i5000</video-mode>
      <consumers>
        <bluefish>
           <device>1</device>
           <sdi-stream>blue_422_channelb</sdi-stream>
           <embedded-audio>true</embedded-audio>
        </bluefish>
      </consumers>
    </channel>
    <channel>
      <video-mode>1080i5000</video-mode>
      <consumers>
        <bluefish>
           <device>1</device>
           <sdi-stream>blue_422_channelc</sdi-stream>
           <embedded-audio>true</embedded-audio>
        </bluefish>
      </consumers>
    </channel>
  </channels>
```

HellGore commented:
> Selection of what output channel and type to use, ( ie. 4224, vs 422 )

4224 here is what is known as external key in CasparCG.

> Support for onboard keyer, if supported in hardware.

I am guessing this is what we know as internal key in CasparCG.

If my above statements are correct perhaps we should use a similar terminology in the bluefish configuration to avoid confusion. Maybe something like:

```
        <bluefish>
           <device>1</device>
           <sdi-stream>a</sdi-stream>
           <keyer>external</keyer>
           <embedded-audio>true</embedded-audio>
        </bluefish>
```

...to indicate what currently would be:

```
        <bluefish>
           <device>1</device>
           <sdi-stream>blue_4224_channela</sdi-stream>
           <embedded-audio>true</embedded-audio>
        </bluefish>
```

jamesdwise commented:
Sorry, @HellGore  I'm a bit confused by the changes you suggested to the syntax we use to configure the channels.  Could you possibly give you suggestion of how this configuration would end up looking

> <channel>
      <video-mode>1080i5000</video-mode>
      <consumers>
        <bluefish>
           <device>1</device>
           <sdi-stream>blue_4224_channela</sdi-stream>
           <embedded-audio>true</embedded-audio>
        </bluefish>
      </consumers>
    </channel>
    <channel>
      <video-mode>1080i5000</video-mode>
      <consumers>
        <bluefish>
           <device>1</device>
           <sdi-stream>blue_422_channelc</sdi-stream>
           <embedded-audio>true</embedded-audio>
        </bluefish>
      </consumers>
    </channel>


Are you suggesting that we use a, b, c, or d to indicate the desired channel?
and the keyer argument to determine if the channel is 422 vs 4224?

Otherwise I am working at integrating most of your other suggestions.
Thank you for being patient with me, this is my first time contributing to such a large project via github.

jesperstarkar commented:
If a-c are the outputs of the card, shouldn't that map to device-number by caspar convention?

jamesdwise commented:
maybe a quick high level overview first.

I'm not sure how familiar everyone is with the Bluefish444 hardware so perhaps it's best if I explain a bit about the hardware first.  Each bluefish device can have up to 4 inputs or 4 outputs, depending on
the hardware and what firmware is loaded.  All devices with an output channels support RGBA and are capable of doing 4224, ie. 2 SDI streams,  Fill and Key, from a single channel.
A small exception exists, that on cards with multiple output channels,
only channels A and C, can actually be configured to do the 4224 output.
(This is more a routing restriction that anything else)
Additionally some of the hardware supports a dedicated onboard hardware keyer. The Keyer in our instance works by taking the SDI input (Background) and combining this on the Bluefish444 hardware with the RGBA data (foreground) on the output channel (supplied by CasparCG). Some customers only require the 422 composited image and other require the composite and the Key data on the second SDI output.

Given that it is possible to have multiple bluefish devices in a machine,
I felt that it was appropriate to use the device number to indicate which
card / device is being used.  I think that using a <sdi-stream> a
[a|b|c|d] </sdi-stream>
is probably the best way to allow users to select which output
channel/stream they want to use.

I am unsure about the best way to provide users with the best balance of
simplicity and functionality from the rest of the config options.
My initial suggestions of the [ blue_4224_channela | blue_422_channela ]
combines the 422 vs 4224 option with the channel selection, and clearly illustrates that only certain hardware channels are capable of 4224 operation.
It also nicely reflects the way the hardware actually works, which makes
it simple for me to implement.   However I see that the simplicity of [a|b|c|d] for channel selection
would be very nice.  

I am open to suggestions of how to use the remaining options to offer
the functionality and flexibility we need.  One possibility would be to use the <keyer> tag with a number of options that allow us to configure, 422, 4224, and Hardware Keyer use.
Although combining all of these may be a bit confusing.
Could you maybe explain the meanings within CasparCG of the <Keyer> function and the arguments that you can use. For example "External" means that you will output a fill channel and a Key channel (4224) to be keyed externally. Or does this mean something else?

I do think that keeping the <hardware-keyer-audio-source>videooutputchannel
[videooutputchannel|sdivideoinput]</hardware-keyer-audio-source>
option is the correct way to go, but again i am open to changing the
actual text and value names here.

I have no Blackmagic hardware to play with so I am unfamiliar with what
the <keyer> options do on their hardware, perhaps
someone who has experience with decklink hardware can enlighten me on
what the various <keyer> options do?

HellGore commented on modules/bluefish/util/blue_velvet.h:
```diff
@@ -33,25 +32,196 @@
+	std::shared_ptr<void>	bvc_;
+
+//BlueVelvetC function pointers members
+	pFunc_bfcGetVersion bfcGetVersion;
```
```
pFunc_bfcGetVersion bfcGetVersion = nullptr;
...etc
```

instead of doing all the assignments in `init_function_pointers()`

HellGore commented:
The Decklink cards are a bit different in that each separately addressable output on a card shows up as a separate device to the application so there each `<device />` in `casparcg.config` can be from the same physical PCIe card, but can also be from separate physical PCIe cards. I think that you should keep the meaning of `<device />` meaning actual physical card as you have done, because it better maps to your hardware.

My only suggestion is to separate the configuration about which output to use and whether to enable external key (4224 output), internal key (hardware keyer in your terminology) or no keying at all.

> Could you possibly give you suggestion of how this configuration would end up looking
> ```
>   <video-mode>1080i5000</video-mode>
>   <consumers>
>     <bluefish>
>        <device>1</device>
>        <sdi-stream>blue_4224_channela</sdi-stream>
>        <embedded-audio>true</embedded-audio>
>     </bluefish>
>   </consumers>
> </channel>
> <channel>
>   <video-mode>1080i5000</video-mode>
>   <consumers>
>     <bluefish>
>        <device>1</device>
>        <sdi-stream>blue_422_channelc</sdi-stream>
>        <embedded-audio>true</embedded-audio>
>     </bluefish>
>   </consumers>
> </channel>
> ```

Certainly:

```
  <video-mode>1080i5000</video-mode>
  <consumers>
    <bluefish>
       <device>1</device>
       <sdi-stream>a</sdi-stream>
       <keyer>external</keyer>
       <embedded-audio>true</embedded-audio>
    </bluefish>
  </consumers>
</channel>
<channel>
  <video-mode>1080i5000</video-mode>
  <consumers>
    <bluefish>
       <device>1</device>
       <sdi-stream>c</sdi-stream>
       <keyer>none</keyer>
       <embedded-audio>true</embedded-audio>
    </bluefish>
  </consumers>
</channel>
```

`<keyer />` should default to `external` when it is supported I think, because that is what most people expect.

HellGore commented on modules/CMakeLists.txt:
```diff
@@ -19,5 +19,4 @@ if (MSVC)
 endif ()
 
 add_subdirectory(image)
-add_subdirectory(reroute)
```
Should not be removed just included at the position it was before.

HellGore commented on modules/bluefish/util/blue_velvet.cpp:
```diff
@@ -23,224 +23,266 @@
+		bvc_ = nullptr;
+		h_module_ = nullptr;
+		init_function_pointers();
+		bvc_ = std::shared_ptr<void>(bfcFactory(), bfcDestroy);
```
Access violation occurs here because if `init_function_pointers()` returns `false` `bfcFactory` will be `nullptr`. You should instead throw the exception `if (!init_function_pointers())`. But also if `bfcFactory()` returns `nullptr`.

HellGore commented on modules/bluefish/util/blue_velvet.cpp:
```diff
@@ -23,224 +23,266 @@
-		InitFunctionsPointers();
-		if (bfcFactory)
-			bvc = bfcFactory();
+		bvc_ = nullptr;
```
A default constructed `std::shared_ptr` is already `nullptr`. Goes for `h_module` as well

HellGore commented on modules/bluefish/util/blue_velvet.cpp:
```diff
@@ -23,224 +23,266 @@
 	{
-		if (bvc)
-			return (BLUE_UINT32)bfcQueryCardProperty32(bvc, iProperty, nValue);
+		if (bvc_)
```
`bvc_` is never `nullptr` after the changes you have done in the constructor.

HellGore commented on modules/bluefish/bluefish.cpp:
```diff
@@ -75,41 +75,25 @@ std::vector<std::wstring> device_list()
-				blue.Detach();
-			}
-		}	
+		bvc_wrapper blue = bvc_wrapper();
```
```
bvc_wrapper blue;
```

HellGore commented on modules/bluefish/bluefish.cpp:
```diff
@@ -57,15 +57,15 @@ std::vector<std::wstring> device_list()
 	try
 	{		
-		BvcWrapper blue = BvcWrapper();
+		bvc_wrapper blue = bvc_wrapper();
```
```
bvc_wrapper blue;
```

HellGore commented on modules/bluefish/bluefish.cpp:
```diff
@@ -42,8 +42,8 @@ std::wstring version()
 	{
-		BvcWrapper blue = BvcWrapper();
-		return u16(blue.GetVersion());
+		bvc_wrapper blue = bvc_wrapper();
```
```
bvc_wrapper blue;
```

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -127,7 +127,7 @@ bool get_videooutput_channel_routing_info_from_streamid(bluefish_hardware_output
 struct bluefish_consumer : boost::noncopyable
 {
-	spl::shared_ptr<BvcWrapper>							blue_;
+	spl::shared_ptr<bvc_wrapper>							blue_;
```
Maybe this should be held by value instead which would also remove the need for `create_blue()` entirely.

HellGore commented:
Could you also comment on my problem with performance? Has that been fixed in your recent commits?

jamesdwise commented:
I Have updated the config file and options to look like ...

>  <bluefish>
                <device>[1..]</device>
		            <sdi-stream>a[a|b|c|d] </sdi-stream>
                <embedded-audio>false [true|false]</embedded-audio>
                <channel-layout>stereo [mono|stereo|matrix|film|smpte|ebu_r123_8a|ebu_r123_8b|8ch|16ch]</channel-layout>
                <key-only>false [true|false]</key-only>
                <keyer>disabled [external|internal|disabled]</keyer>
                <internal-keyer-audio-source> videooutputchannel [videooutputchannel|sdivideoinput]</internal-keyer-audio-source>
            </bluefish>

Where, disabled gives 422 output, external gives a 4224 configuration, and internal uses the dedicated keyer on the bluefish hardware.  I have also updated the keyer audio source to better reflect that it is only for the internal keyer.
Also the sdi-stream selection is now just a,b,c,d, which makes thing s a lot clearer.

I do have a questions though...
Where is the correct place to tell users  about specific hardware restrictions, ie. the external keyer mode will only work on ch a or c?  

Lastly I have seen the performance issues you mentioned and i am looking into it.

HellGore commented:
> Where is the correct place to tell users about specific hardware restrictions, ie. the external keyer mode will only work on ch a or c?

```
<sdi-stream>a[a|b|c|d] (b and d never supports external key)</sdi-stream>
```

...or

```
<keyer>disabled [external|internal|disabled] (external never supported when sdi-stream is b or d)</keyer>
```

Or if you want to use the wording "only supported with a and c", but c is not supported on the card I have here for testing if I understand correctly because it only has 3 BNC in total.

jamesdwise commented:
Hey all,

So I am still looking into the performance issue,  and it appears (for me anyway) to go
away when I dont attach any screen consumers to the channels.
ie. only have bluefish consumers on each channel.

I am looking at a few different mechanisms to improve the performance inside the bluefish code as well.
However, it looks like caspar, despite launching the send_frame func inside an executor, 
is still waiting on some results from bluefish.     am trying to understand this mechanism better, 
but if anyone would like to share some knowledge that would be most appreciated.  
Eg. If i reduce the size of the video buffer that gets DMA'd to say 1/4 of its' original size,
everything works fine with 3 channels, of both screen and bluefish consumers.
indicating that caspar is still waiting on some result from the bluefish consumer?
Shouldn't caspar be insulated from the length of time of my DMA op? 


Additionally:
On some systems, I also get a large number of 
" [buffer] Performance warning. Buffer mapping blocked: 0.022"
messages, I am assuming this out out of my control ( or at least not related to bluefish)
and that dropping frames where i get this warning is acceptable?

HellGore commented:
CasparCG sends the frame to each consumer on a channel via the `send()` method and then waits for all `std::future<bool>` to return a result. This allows the consumers to keep the rendering pace of the video channel, because CasparCG itself doesn't have access to the black burst, but the consumers like bluefish and decklink do.

The problem here seems to be that sporadically the `bluefish_consumer::display_frame()` seems to take two frame durations to complete which leaves me to believe that it has something to do with `bfcWaitVideoOutputSync()` perhaps it should be done at the end of the method instead of the beginning or something.

I have the bluefish SDK on my hard drive but I can't find any API documentation about `bfcWaitVideoOutputSync()` only in examples.

If you write `diag` on the CasparCG command line you can see clearly in the graph that each bluefish consumer instance produces that when the performance problem occurs it seems like the red `sync-time` goes up to exactly a whole frame duration. `frame-time` is also close to an entire frame duration most of the time which is a bit alarming.

Theoretically `sync_time` plus `frame_time` should always add up to one frame duration, where `sync_time` is the "sleep" time needed to get the correct channel pace.

HellGore commented on modules/bluefish/util/blue_velvet.cpp:
```diff
@@ -20,67 +20,205 @@
-}
+	BLUE_UINT32 bvc_wrapper::attach(int iDeviceId)
+	{
+		return bfcAttach((BLUEVELVETC_HANDLE)bvc_.get(), iDeviceId);
```
None of these casts to `BLUEVELVETC_HANDLE` should be needed since `bvc_.get()` returns `void*` anyway.

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -277,32 +475,31 @@ struct bluefish_consumer : boost::noncopyable
-											nullptr,                 
-											BlueImage_DMABuffer(reserved_frames_.front()->id(), BLUE_DATA_IMAGE));
+											BlueImage_DMABuffer(reserved_frames_.front()->id(), BLUE_DATA_IMAGE),
+											0);
```
It seems to be this `bfcSystemBufferWriteAsync()` call that takes almost linearly more time when more SDI outputs are used at the same time on the same card. So when I run 3 simultaneous channels this call takes 3 times as long on each channel as it does when only using one channel. They seem to be contending for some resource.

What does the `OVERLAPPED* pOverlap` parameter do? It is always set to `nullptr`. Does that mean that it is not asynchronous but blocking?

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -277,32 +475,31 @@ struct bluefish_consumer : boost::noncopyable
-											BlueImage_HANC_DMABuffer(reserved_frames_.front()->id(), BLUE_DATA_IMAGE));
+			blue_->system_buffer_write(const_cast<uint8_t*>(reserved_frames_.front()->image_data()), 
+											static_cast<unsigned long>(reserved_frames_.front()->image_size()),  
+											BlueImage_HANC_DMABuffer(reserved_frames_.front()->id(), BLUE_DATA_IMAGE), 
```
Should `BlueImage_HANC_DMABuffer` be replaced with `BlueImage_HANC_DMABuffer` here as well or is it correct even though it is about video not sound?

jamesdwise commented:
Lots of good Q.'s there, 
I will try and answer them all.

1. The `BlueImage_HANC_DMABuffer`  usage is correct, we use the macro to say what buffers we are dealing with, and then the argument, ie. `BLUE_DATA_IMAGE` or `BLUE_DATA_HANC ` to indicate exactly what sort of DMA transfer is happening.
You will notice that in the case where we are not playing out audio/hanc we use the `BlueImage_DMABuffer` macro instead.

2. `bfcSystemBufferWriteAsync()`  is our DMA call, we only have a single DMA engine on the card, so it makes sense that the call takes up to 3 times as long to return, when you are doing 3 transfers.  Our driver allows us to make multiple calls from multiple threads, but due to the limitation of the hardware,  only 1 ever executes at a time. So the first call will return in the actual DMA time, but if you make 3 calls in close succession, then the 3rd one will only return once all 3 have completed. 
The `OVERLAPPED* pOverlap`  argument can be used for asynchronous transfers, by passing a `nullptr `  we force the operation to happen synchronously and blocking.  Given that the the send_frame happens on its own executor and that we still need to wait for the DMA to complete before making the present call, I didn't think it was worth the effort to make this a asynchronous operation.   - This is one of the performance improvements I am testing now

3. I have been using the `diag` tool already, but thank you for the explanation regarding what the `frame-time` and `sync-time` values are supposed to mean, it is always better than guessing.
`bfcWaitVideoOutputSync()`  is our basic Wait for sync operation it is the equivalent of the old `wait_output_video_synch` call.

I am currently looking at a couple of different ways to improve the performance,
- changing the DMA to happen Asynchronously
- modifying the code to our card in a `FIFO `mode instead of `FRAMESTORE`  which adds some latency but also buffering to the entire process.
- disconnecting the DMA from the critical path and returning to caspar asap.

Keep in mind our current hardware is only PCI-E  Gen 1 x4 lane, so the max possible bandwidth you can get to the card is between 700 and 800MB/s.  3 streams of 1080i50 RGBA is around 650MB/s  so whilst possible we are already close to the DMA limits of what is possible.  This means that when doing 3 streams we require almost the entire 40ms (actually around 35ms) available to transfer 3 channels worth  of data.   If we get the first of those frames even  5ms after the last wait for sync call completes there will always be a dropped frame.   Hopefully this provides some insight as to why you might be seeing the performance issues you are.   Of course adding some buffering and async support should help alleviate these issues.

HellGore commented:
It seems to me that the problem might be that calling `bfcWaitVideoOutputSync()` in times when waiting is unnecessary causing one extra frame duration to pass. Like for example if 39ms has passed since the last frame was shown and we call the `bfcWaitVideoOutputSync()` method the actual wait might slip over to the next frame duration causing 41ms to pass instead of 1ms. Is this something that might be the cause of the hickups?

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -196,30 +251,177 @@ struct bluefish_consumer : boost::noncopyable
+				keyer_control_value = VIDEO_ONBOARD_KEYER_SET_STATUS_ENABLE_OVER_BLACK(keyer_control_value);
+			else
+				keyer_control_value = VIDEO_ONBOARD_KEYER_SET_STATUS_DISABLE_OVER_BLACK(keyer_control_value);
+		
```
I have been playing with the Neutron 1i20 firmware a little bit now instead of 3o and noticed that the internal keyer expects RGB channels to be straight non multiplied. CasparCG's native output mode is to premultiply RGB channels with alpha. perhaps the `VIDEO_ONBOARD_KEYER_SET_STATUS_DATA_IS_PREMULTIPLIED` mask is needed as well?

HellGore commented:
> Also minor change to tell the Keyer that data is pre-multiplied when we receive it.

I don't see any changes that solves this problem. It is still multiplied twice, making the image darker when alpha is between 0.0 and 1.0

HellGore commented:
Now that the actual delay between a frame arriving at the consumer and it being presented on the SDI output has increased you will need to change the return value of `bluefish_consumer_proxy::buffer_depth()` accordingly. The `buffer_depth` of a consumer is how many frames worth of delay that the consumer itself adds. It is used by the `output` class in CasparCG to try to make all consumers play in sync with regards to presentation to eye and ears.

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -237,15 +455,85 @@ struct bluefish_consumer : boost::noncopyable
+		wait_b.attach(blue->device_index_);
+		EBlueVideoChannel out_vid_channel = get_bluesdk_videochannel_from_streamid(blue->device_output_channel_);
+		wait_b.set_card_property32(DEFAULT_VIDEO_OUTPUT_CHANNEL, out_vid_channel);
+		int frames_to_buffer = 3;
```
I tried 1 here and I still got no dropped frames when running the 3o firmware with SDI A, B and C. Did you see a need for more than 1?

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -75,7 +133,14 @@ struct bluefish_consumer : boost::noncopyable
 
-	std::array<blue_dma_buffer_ptr, 4>					reserved_frames_;	
+	std::array<blue_dma_buffer_ptr, 4>					all_frames_;	
+	std::queue<blue_dma_buffer_ptr>						reserved_frames_;
```
I think it would be much simpler and less error prone to use `tbb::concurrent_queue` or `tbb::concurrent_bounded_queue` for both `reserved_frames_` and `live_frames_`. With no need for manual mutex handling.

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -75,7 +133,14 @@ struct bluefish_consumer : boost::noncopyable
+	std::queue<blue_dma_buffer_ptr>						live_frames_;
+	std::mutex											live_frames_lock_;
+	std::shared_ptr<std::thread>						dma_present_thread_;
+	bool												end_dma_thread_;
```
This needs to be a `tbb::atomic<bool>` to be able to be guarantee that the change from `false` to `true` is memory visible by the DMA thread, otherwise a cached value of `false` could be read.

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -237,15 +455,85 @@ struct bluefish_consumer : boost::noncopyable
+
+		while (!blue->end_dma_thread_)
+		{
+			if (blue->live_frames_.size() && BLUE_OK(blue->blue_->video_playback_allocate(buffer_id, underrun)))
```
Calling `blue->live_frames_.size()` without lock on `live_frames_lock_` is not correct, right? Anyway, with the change to TBB queues instead this can be done in a different way.

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -254,56 +542,59 @@ struct bluefish_consumer : boost::noncopyable
+			if (dma_present_thread_ == 0)
+			{
+				end_dma_thread_ = false;
+				dma_present_thread_ = std::make_shared<std::thread>(&dma_present_thread_actual, this);
```
Should be

```
dma_present_thread_ = std::make_shared<std::thread>([this]{dma_present_thread_actual();});
```

...is better because then you can change the `dma_present_thread_actual` function to be an instance member function instead.

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -237,15 +455,85 @@ struct bluefish_consumer : boost::noncopyable
 	}
 
-	void display_frame(core::const_frame frame)
+	static void dma_present_thread_actual(void* arg)
```
Was there ever a need for this argument to be `void*` instead of `bluefish_consumer*`? Anyway with the changes I suggested regarding the thread constructor this signature will be changed to `void dma_present_thread_actual()` with ordinary implicit `this` pointer.

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -397,6 +401,10 @@ struct bluefish_consumer : boost::noncopyable
 				CASPAR_THROW_EXCEPTION(caspar_exception() << msg_info(" Failed to get invalid video mode flag"));
 
+			// The bluefish HW keyer is going to pre-multiply the RGB with the A. 
+			// This setting results in the correct image coming through when using the 1080_test.tga image.
```
It should really be `VIDEO_ONBOARD_KEYER_SET_STATUS_DATA_IS_PREMULTIPLIED` for CasparCG. Notice `IS` instead of `IS_NOT`.

>Hardware Keyer:
>Now using the correct Macro to setup the hw Keyer,
>Be aware the terminology can be confusing as the the macro refers to the source data, not the process to be applied.

The source data **is** premultiplied.

>This has been tested with the casper 1080_test.tga file and produces the correct output.
>Ie. we see the following text in white.
>"Alpha / Key correct: Premultipled
> Linear / Additive"

Unfortunately that file seems to be non useful for this test.

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -237,15 +463,77 @@ struct bluefish_consumer : boost::noncopyable
+
+		while (!end_dma_thread_)
+		{
+			if (!live_frames_.empty() && BLUE_OK(blue_->video_playback_allocate(buffer_id, underrun)))
```
Actually I am not sure if `tbb::concurrent_bounded_queue::empty` is guaranteed to give the expected result. It is based on `size()` and that doesn't seem that convincing in the api docs:
> Number of pushes minus number of pops. The result is negative if there are pop operations waiting for corresponding pushes. The result can exceed capacity() if the queue is full and there are push operations waiting for corresponding pops."

The correct way is to use the `try_pop()` method, which pops if an item is available and returns `true` otherwise `false`. You can see example uses of that method in other places in the CasparCG Server code.

HellGore commented on modules/bluefish/consumer/bluefish_consumer.cpp:
```diff
@@ -254,56 +542,54 @@ struct bluefish_consumer : boost::noncopyable
 		// Copy to local buffers
-		
-		if(!frame.image_data().empty())
+		if (!reserved_frames_.empty())
```
Same thing here with `try_pop()` instead

HellGore commented:
I think this is shaping up nicely and with the few changes I proposed now I feel that it is almost ready.

Btw, how simple is it to add Linux support? The `dlopen/dlsym` stuff would be simple enough, is there any other things in your Linux SDK that differs much?

jamesdwise commented:
I have now applied to latest changes as per your comments.

Given the confusion I seem to be having with the correct HW keyer setting, and that apparently the test file I was using does not produce the correct output, could you possibly suggest a test that would help me to ensure i have the keyer setting correct?

As for Linux/otherOS support i was initially planning to do OSX support before Linux, but i have seen that the OSX build is not really in good working [order.....](https://github.com/CasparCG/Server/issues/428)
So perhaps i will focus on the Linux support first.
Adding Linux support should be pretty easy, we have a BlueVelvetC impl available on linux so it is just a matter of assigning all the function pointers inside the wrapper class, and everything should be good to go.

However in either case I was planning to add producer support before adding Other OS support.

HellGore commented:
A question I have is that when I exit CasparCG solid black is sent instead of fully transparent so when using the internal keyer I see whatever was playing in the foreground at exit fully opaque instead of only the background. The same happens with external key but then it is not full opaque but instead the last transparency when exiting CasparCG.

Is there something I can do in `BlueProps.exe` to affect what is being output when an application is not using the card? Or is the application responsible of showing transparent before exiting?

HellGore commented:
> Given the confusion I seem to be having with the correct HW keyer setting, and that apparently the test file I was using does not produce the correct output, could you possibly suggest a test that would help me to ensure i have the keyer setting correct?

Certainly! Given this being played as background on the SDI input:

![background](https://cloud.githubusercontent.com/assets/2980939/22977163/da4449d8-f38d-11e6-9c78-efb2a6cf83a4.png)

...and this being played through CasparCG as the foreground using internal keyer (note seems completely white in browser but has non opaque alpha):

![alpha_gradient](https://cloud.githubusercontent.com/assets/2980939/22977194/f167d01c-f38d-11e6-86ca-2c335bdbf284.png)

...the result should look like:

![correct_compositing](https://cloud.githubusercontent.com/assets/2980939/22977202/fbcf2a14-f38d-11e6-99e6-022b132594c3.png)

If however the RGB values are multiplied twice with the A, which essentially is what happens when this is done incorrectly this happens:

![incorrect_compositing](https://cloud.githubusercontent.com/assets/2980939/22977269/309bac5e-f38e-11e6-9bb6-0a7fff439f1c.png)

HellGore commented:
> As for Linux/otherOS support i was initially planning to do OSX support before Linux, but i have seen that the OSX build is not really in good working order.....
So perhaps i will focus on the Linux support first.

Yes, definately Linux support first. An OSX port of CasparCG is currently not on the road map at all.

jamesdwise commented:
Thanks for the Info about testing the various keyer settings.

As for the state of the device when exiting caspar, you can use BlueProps.exe to enable the bypass relay(s)
that will forcibly pass the video from the input, through to the input, without any interaction with the card.

Additionally when we exit caspar we display a blackframe, via the VIDEO_BLACKGENERATOR property,
which i think explains why you are seeing the results you when configuring the keyer in external mode.
Assuming that you are using an external keyer, us generating black on fill and key, will result in your external keyer displaying the background video.

This is different to when configuring the keyer as internal, we still turn the black Generator on, but this happens at the end of the chain, therefore blocking the incoming SDI signal.

Does the above explanation make sense?
If you have recommendations regarding the correct behavior on exit
I am open to modifying the code,  

One possible solution would be to,  DMA a black frame upon exit, instead of enabling the VIDEO_BLACKGENERATOR, this would result in the input being displayed on the input when internal keyer is set on, and the same / current behavior should be maintained when using an external keyer.

(upon exit the card will automatically cycle the last buffer displayed, but no audio)

Hope this helps..

HellGore commented:
> Additionally when we exit caspar we display a blackframe, via the VIDEO_BLACKGENERATOR property,
which i think explains why you are seeing the results you when configuring the keyer in external mode.
Assuming that you are using an external keyer, us generating black on fill and key, will result in your external keyer displaying the background video.

But it is not displaying black on the key SDI output, but rather the last transparency used when playback was active in caspar. So if i for example played a fully opaque `PLAY 1-10 RED` before exiting, the FILL SDI indeed goes black (i guess that is the `VIDEO_BLACKGENERATOR` property), but it actually still shows full white on the KEY SDI signal so the background will be blocked when the downstream keying is performed on the mixer. My suggestion here is to ensure that the KEY SDI also goes black, if that is possible.

jamesdwise commented:
Hey All,

Just wondering if there is anything else required from me for this pull request to be accepted?
Would you prefer I withdraw the original request and make a new one?
Or is there some more changes that you would like me to make before accepting the pull request?


- James

(hope i'm not being pushy here, i just to keep progressing)

HellGore commented:
No, don't withdraw it and make a new one. I will have a look on monday again. I have been busy with other stuff.

jesperstarkar commented:
Applause

sfahadshahzad commented:
any update regarding OSX support.

Fahad

HellGore commented:
Do you mean Linux?

HellGore commented:
There is no planned support for Max OS in CasparCG.

jesperstarkar commented:
@jamesdwise I'm testing Bluefish for the first time, and about to write a summary and send back to Bluefish about my experiences.

Would you be available/interested in implementing further on this? I'm thinking about a couple of things needed like reporting cards and capabilities in `INFO SYSTEM` calls.

I'm also thinking that a device with a given firmware/configuration should be internally mapped as different devices in CasparCG, so that a `INFO SYSTEM` would report the number of available channels/devices/streams available. This also affects how your configuration between devices and streams in the config file looks. 

Please let me know if your up for further discussions and potentially putting down some more work on this.

jamesdwise commented:
@jesperstarkar  Nice to hear your using bluefish, we would definitely be intersted to hear feedback regarding your experiences.

I am interested in doing more work on this, primarily input/ producer support and Linux support.
In fact I have already added input/producer support, and the code is currently on my github page, 
I am just doing some more testing before I make a pull request to get it pulled into the main branch.

Feel free to grab the branch from my page and test the input support as well.

As for you suggestions:
- Yes adding information about each devices capabilities would be a good idea,

- Currently each card/device is already addressed via a different deviceID in the config file already
(different to how the blackmagic cards work)   so i'm not exactly sure what you mean here?

Happy to discuss,
  James

jesperstarkar commented:
Hi James. Sorry about being a bit vague, I just wanted to get in contact and hear if you still invested time and development effort into this. 

I'll make a separate thread for discussing devices vs streams.

jesperstarkar commented:
Btw James - what are you thinking about implementing a proper watch dog. Not only tied to the CasparCG process, but relying on constantly getting pings from CasparCG to keep alive. Such a watch dog would enable relay bypass in case of software hangups. It would be a big advantage and sales argument for Bluefish over Decklinks - as they don't support that kind of relays.

jamesdwise commented:
Hi Jesper,
Implementing a proper watch dog might be possible, 
Obviously this will only be useful for people using input and output firmwares and/or the keyer.
Clearly this will be more useful once i have added the input/producer support.

We would need to configure this via an argument in the config file, and also include a setting to indicate the length of time to set before the bypass relay(s) is enabled. 

Eg. Adding a...
`<bypass> 40</bypass>`
to the config setting for a Bluefish channel, could configure things such that after 40ms of no response the card would automatically turn on the bypass relays.
a setting of 0 could or omitting the setting could be used to NOT set any watchdog.

I'm currently trying to think through the various workflow possibilities of this feature, so feel free to add in your expectations and thoughts.
Eg. What happens when Caspar crashes, and then the user wants to restart, but the bypass relay(s) have been enabled? When should we transition back to displaying the caspar output? ie. disable the bypass relay(s)

jamesdwise commented:
I have been looking at the possibility of a hw watchdog.
it is definitely possible.  However the length of time before it enables the bypass relay(s)
is measured in interrupts, not ms.
Would setting this to a constant of 2 missed interrupts initiates the watchdog be suitable?
or do you really want it to be a configurable amount?

jesperstarkar commented:
When it comes to going back from a bypass to CasparCG output, I think it should happen instantaneously, with no configuration.

I don't know what the correct number of interrupts in the threshold should be. It isn't hard to make this a config variable with `2` as the default value, is it? What is the largest number you accept?

jamesdwise commented:
I think that 2 would be the minimum, and the default.
I'm not even sure why you would want a larger number,
but we can support values up to at least into the thousands...

I think that if added it should be enabled by default,
and a special argument in the config file can be used to change the default number of interrupts to wait, OR to turn it off, eg. <hw-watchdog> <0>  using 0 to indicate we don't want it to run.

I'm not sure how open the casparCG community is to accepting custom config commands?

jesperstarkar commented:
I can only speak for myself. I mean, this value has a default, so it is opt-in to use the new parameter. And adding a new parameter either to a top-level bluefish-tag, or as a child of the consumer, will be fully backwards compatible. I don't think there's any resistance again adding a new parameter.
